plugins/koterm.koplugin/main.lua

```diff
@@ -0,0 +1,561 @@
+            print("Maybe an incorrect shell: '" .. shell .. "'\n")
+            os.exit()
+        end
+        os.exit()
+        return
+    end
+
+    self.is_shell_open = true
+    if Device:isAndroid() then
+        -- feed the following commands to the running shell
+        self:transmit(" export TERM=vt52\n")
+        self:transmit(" stty cols " .. cols .. " rows " .. rows .."\n")
+    end
+
+    self.input_widget:resize(rows, cols)
+    self.input_widget:interpretAnsiSeq(self:receive())
+
+    logger.info("KOTerm: spawn done")
+end
+
+function Terminal:receive()
+    local last_result = ""
+    repeat
+        C.tcdrain(self.ptmx)
+        local count = tonumber(C.read(self.ptmx, self.chunk, self.chunk_size))
+        if count > 0 then
+            last_result = last_result .. string.sub(ffi.string(self.chunk), 1, count)
+        end
+    until count <= 0 or #last_result >= self.chunk_size - 1
+    return last_result
+end
+
+function Terminal:refresh(reset)
+    if reset then
+        self.refresh_time = 1/32
+        UIManager:unschedule(Terminal.refresh)
+    end
+
+    local next_text = self:receive()
+    if next_text ~= "" then
+        self.input_widget:interpretAnsiSeq(next_text)
+        if #self.input_dialog:getInputText() > self.buffer_size then
+            local input = self.input_dialog:getInputText()
+            input = input:sub(#input - self.buffer_size)
+            self.input_dialog:setInputText(input)
+        end
+        if self.is_shell_open then
+            UIManager:tickAfterNext(function()
+                UIManager:scheduleIn(self.refresh_time, Terminal.refresh, self)
+            end)
+        end
+    else
+        if self.is_shell_open then
+            if self.refresh_time > 5 then
+                self.refresh_time = self.refresh_time
+            elseif self.refresh_time > 1 then
+                self.refresh_time = self.refresh_time * 1.1
+            else
+                self.refresh_time = self.refresh_time * 2
+            end
+            UIManager:scheduleIn(self.refresh_time, Terminal.refresh, self)
+        end
+    end
+end
+
+
+function Terminal:transmit(chars)
+    C.write(self.ptmx, chars, #chars)
+    self:refresh(true)
+end
+
+
+--- kills a running shell
+-- @param ask if true ask if a shell is running, don't kill
+-- @return pid if shell is running, -1 otherwise
+function Terminal:killShell(ask)
+    UIManager:unschedule(Terminal.refresh)
+    local pid_file = io.open("koterm.pid", "r")
+    if not pid_file then
+        return -1
+    end
+
+    local pid = tonumber(pid_file:read("*a"))
+    pid_file:close()
+
+    if ask then
+        return pid
+    else
+        local terminate = "\03\n\nexit\n"
+        self:transmit(terminate)
+        -- do other things before killing first
+        self.is_shell_open = false
+        self.history = ""
+        os.remove("koterm.pid")  -- check this todo
+        C.close(self.ptmx)
+
+        C.kill(pid, C.SIGTERM)
```

Comment:
I noticed a few (sh) zombies after launching and quit'int a few times.
You need to C.waitpid() the pid after killing it, see https://github.com/koreader/koreader-base/blob/ab2fa6f3aedb3e9a799a10df39b3bd5346de3f14/ffi/util.lua#L370-L400